Deprecate HanaDB, HanaTranslator and update example notebook to use new implementation

yberber-sap opened the pull request:
This PR includes two main changes:

1. **Deprecation of `HanaDB` Vector Store and `HanaTranslator`**
   - Added `@deprecated` annotation to the both classes.
   - Set `since="0.1.0"` and `removal="1.0"` to indicate the deprecation timeline. (without removal, it fails)
   - Suggested using the updated implementation from the `langchain_hana` package.
   - Added link to the official SAP integration repository:
     https://github.com/SAP/langchain-integration-for-sap-hana-cloud

2. **Notebook Update**
   - Adapted the example notebook to use the new `HanaDB` class from `langchain_hana`.
   - Ensured that usage is aligned with the new API and recommended practices.
   - Added examples for the new features

Closes #<issue-number-if-any>.

yonarw commented on libs/community/langchain_community/vectorstores/hanavector.py:
```diff
@@ -418,14 +437,13 @@ def similarity_search_with_score_and_vector_by_vector(
             f'  TO_NVARCHAR("{self.vector_column}"), '  # row[2]
             f'  {distance_func_name}("{self.vector_column}", TO_REAL_VECTOR '
-            f"     (ARRAY({embedding_as_str}))) AS CS "  # row[3]
+            f"     ('{str(embedding)}')) AS CS "  # row[3]
```
I compared this to the way Xingqi changed it in the [hnsw_index pull request](https://github.com/langchain-ai/langchain/pull/27884/files#diff-e8b245923954c1cbed13f1508c3ece266df0a1d886f4680505e352c746335334R504). 
I think your version is cleaner. Could you change the TO_REAL_VECTOR usage in the other places of this file as well . To not get merge conflicts with Xingqis change you might need to rebase your changes ontop of hers. Hopefully the PR will get merged at some point.

yonarw commented on libs/community/tests/integration_tests/vectorstores/test_hanavector.py:
```diff
@@ -1432,3 +1432,80 @@ def test_preexisting_specific_columns_for_returned_metadata_completeness(
     assert docs[0].metadata["quality"] == "good"
     assert docs[0].metadata["ready"]
     assert "NonExisting" not in docs[0].metadata.keys()
+
```
Same here: To avoid merge conflicts, maybe rebase your changes on the [hnsw_index PR.](https://github.com/langchain-ai/langchain/pull/27884/files#diff-e8b245923954c1cbed13f1508c3ece266df0a1d886f4680505e352c746335334R504)

yonarw commented on docs/docs/integrations/vectorstores/sap_hanavector.ipynb:
```diff
@@ -41,11 +41,11 @@
   {
    "cell_type": "code",
-   "execution_count": 11,
+   "execution_count": null,
```
These changes come from running the notebook. I think it might be ok if we also add the output (see comment below).

yberber-sap commented on libs/community/langchain_community/vectorstores/hanavector.py:
```diff
@@ -399,6 +420,85 @@ def similarity_search_with_score(
+        """Extract metadata columns involved in keyword search."""
+        keyword_columns = set()
+
+        def recurse_filters(f, parent_key=None):
```
I am so sorry for the misunderstanding. As I mentioned in the email, the PR request that needs to be reviewed is in the hana-keyword-filtering branch https://github.com/yberber-sap/langchain/pull/2 , where I have resolved almost all these issues. The code in the hana-keyword-filtering branch is almost the same; I just removed unused lines and adapted the tests.

The recurse_filter function has no depth limitations. We use it to find unspecific metadata columns that are used with the $contains operation. If we use the $and operation, the depth might be 3:

filters = {"$and": [{"name": {"$contains": "bob"}}, {"name": {"$contains": "johnson"}}]}

Without $and or $or operations, the depth is 2, such as:

{"name": {"$contains": "bob"}}

I thought it was not necessary to enforce a depth limitation because there is no risk of an infinite loop. However, I could add a depth limit of 3, assuming we do not use $and or $or operations nested within another $and or $or operation (which is probably not allowed).

yberber-sap commented on libs/community/tests/integration_tests/vectorstores/test_hanavector.py:
```diff
@@ -1468,43 +1468,43 @@ def test_hanavector_keyword_search(texts: List[str], metadatas: List[dict]) -> N
-    )
-    assert len(docs) == 1
-    assert "good" in docs[0].metadata["quality"]
+# @pytest.mark.skipif(not hanadb_installed, reason="hanadb not installed")
```
solved at https://github.com/yberber-sap/langchain/pull/2

yberber-sap commented on libs/community/langchain_community/vectorstores/hanavector.py:
```diff
@@ -418,6 +420,85 @@ def similarity_search_with_score(
+                            recurse_filters(subfilter)
+                    else:
+                        recurse_filters(value, parent_key=key)
+            # elif isinstance(f, list):  # Handle lists of filters (logical operator operands)
```
solved at https://github.com/yberber-sap/langchain/pull/2

yberber-sap commented on libs/community/langchain_community/vectorstores/hanavector.py:
```diff
@@ -604,12 +692,15 @@ def _process_filter_object(self, filter):  # type: ignore[no-untyped-def]
+                    where_str += f"SCORE(? IN (\"{key}\" EXACT SEARCH MODE \'text\')) > 0"
+                    query_tuple.append(special_val)
+
+                    # if key  == self.content_column or key in self.specific_metadata_columns:
```
solved at https://github.com/yberber-sap/langchain/pull/2

yberber-sap commented:
I am closing this PR. See https://github.com/yberber-sap/langchain/pull/2

yberber committed:
community[patch]: Add deprecation warnings to HanaDB and HanaTranslator implementations

- Add deprecation warnings to hana implementation that are in langchain-hana.

yberber committed:
docs: update HANA vector store notebook to use langchain-hana HanaDB implementation

- Replaced usage of the deprecated HanaDB class with the new HanaDB from the
  langchain_hana package.
- Updated import paths and example code to reflect the recommended migration path.
- Added examples demonstrating new features such as keyword search and internal embeddings.
